### PR TITLE
0.3 release blog post.

### PR DESCRIPTION
**Pull request by robertnishihara**

### CODE REVIEW

**Comment by atumanov** [comment]
I’m wondering if “distributed actors” is the right name for the feature. This feature essentially implements actor [RMI](https://en.wikipedia.org/wiki/Remote_method_invocation) (remote method invocation). The term distributed actors rather implies that actors themselves are distributed, whereas the intended semantics is that the actor handles can be passed around and actor methods can be invoked remotely via those handles. Would actor RMI be a better term for this?

File: site/_posts/2017-11-27-ray-0.3-release.markdown [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+logger = Logger.remote()
+# Pass the actor handle to some tasks.
+tasks = [task.remote(logger, i) for i in range(10)]
+# Wait for the tasks to finish.
+ray.get(tasks)
+# Get the logs from the tasks.
+print(ray.get(logger.read_logs.remote()))
+```
+
+This feature is still under development, and full functionality is coming soon.
```
Comment (atumanov) [inline site/_posts/2017-11-27-ray-0.3-release.markdown]: It would be good to expand on what's meant by "full functionality". It leaves me to wonder what's missing in the current implementation. One way to address this could be to supplement the feature list below with a list of limitations with a note saying that the latter is under active development/design.

File: site/_posts/2017-11-27-ray-0.3-release.markdown [hunk 2]
```diff
@@ -0,0 +1,158 @@
+---
+layout: post
+title: "Ray: 0.3 Release"
+excerpt: "This post announces the release of Ray 0.3."
+date: 2017-11-29 00:00:00
+---
+
+We are pleased to announce the Ray 0.3 release. This release includes a number
+of bug fixes, as well as [**distributed actors**][1] and a [**hyperparameter
+search library**][2] (including integration with [**RLlib**][3]).
```
Comment (atumanov) [inline site/_posts/2017-11-27-ray-0.3-release.markdown]: I would really emphasize the main release features, particularly RLlib-related. How about : 
This release introduces **Ray.tune** --- the new RLlib-integrated hyperparameter search library, distributed actor RMI, and a number of stability and performance enhancements and bug fixes.

File: site/_posts/2017-11-27-ray-0.3-release.markdown [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+})
+```
+
+In-progress results can be visualized live using tools such as Tensorboard and
+rllab's VisKit (or you can read the JSON format logs directly from the driver
+node):
+
+View the [documentation][2] and the [code][9].
+
+### RLlib integration with Ray.tune
```
Comment (atumanov) [inline site/_posts/2017-11-27-ray-0.3-release.markdown]: would it make sense to swap the order, i.e., **Ray.tune integration with RLlib**

**Comment by ericl** [comment]
How about something like "direct actor calls"? That seems to be the main use case it enables.

File: site/_posts/2017-11-27-ray-0.3-release.markdown [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+```
+
+This feature is still under development, and full functionality is coming soon.
+Currently, the feature behaves as follows.
+
+- Execution order follows submission order, per handle. This means that Ray only
+  guarantees in-order execution between method calls from the same handle. No
+  ordering guarantees are currently provided between methods invoked from
+  different handles to the same actor.
+- Actor replay includes method calls submitted by the actor’s creator through
```
Comment (atumanov) [inline site/_posts/2017-11-27-ray-0.3-release.markdown]: actor replay is not introduced. Would it make sense to mention some initial support for actor fault tolerance here?

File: site/_posts/2017-11-27-ray-0.3-release.markdown [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+# Wait for the tasks to finish.
+ray.get(tasks)
+# Get the logs from the tasks.
+print(ray.get(logger.read_logs.remote()))
+```
+
+This feature is still under development, and full functionality is coming soon.
+Currently, the feature behaves as follows.
+
+- Execution order follows submission order, per handle. This means that Ray only
```
Comment (atumanov) [inline site/_posts/2017-11-27-ray-0.3-release.markdown]: - all actor method invocations are serialized, i.e., all actor methods are reentrant.
- actor method invocation order : actor methods invoked on the same handle are executed in the order of invocation. No ordering guarantees are provided for method calls invoked on different actor handles.

File: site/_posts/2017-11-27-ray-0.3-release.markdown [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+  guarantees in-order execution between method calls from the same handle. No
+  ordering guarantees are currently provided between methods invoked from
+  different handles to the same actor.
+- Actor replay includes method calls submitted by the actor’s creator through
+  the original handle. If the actor is lost, tasks submitted through passed
+  handles may not re-execute when the actor is reconstructed.
+- Distributed actor replay currently does not support checkpointing.
```
Comment (atumanov) [inline site/_posts/2017-11-27-ray-0.3-release.markdown]: I would elaborate on checkpointing here. What exactly does it mean? That actor reconstruction will not use checkpointed state? Or it will not checkpoint during the process of reconstruction? Does the reader know about actor state checkpointing as a feature? This bullet could really be expanded into a couple of sentences.

File: site/_posts/2017-11-27-ray-0.3-release.markdown (hunk 2 above)
Comment (ericl) [reply site/_posts/2017-11-27-ray-0.3-release.markdown]: I'd like to hold off on emphasizing RLlib, until we have the refactoring completed / paper posted, i.e. 0.4

File: site/_posts/2017-11-30-ray-0.3-release.markdown [hunk 7]
```diff
@@ -0,0 +1,147 @@
+---
+layout: post
+title: "Ray: 0.3 Release"
+excerpt: "This post announces the release of Ray 0.3."
+date: 2017-11-30 14:00:00
+---
+
+We are pleased to announce the Ray 0.3 release. This release introduces
+[**Ray.tune**][2] --- a new hyperparameter search library, [**distributed actor
+handles**][1], and a number of stability and performance enhancements and bug
```
Comment (ericl) [inline site/_posts/2017-11-30-ray-0.3-release.markdown]: Nit: That's a lot of ands in a row

File: site/_posts/2017-11-30-ray-0.3-release.markdown [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+```bash
+cd ray/python/ray/rllib
+python train.py -f tuned_examples/cartpole-grid-search-example.yaml
+```
+
+The `tuned_examples` directory also contains pre-tuned hyperparameter
+configurations for common benchmark tasks such as Pong and Humanoid. View the
+[RLlib documentation][3].
+
+### Initial PyTorch support for RLlib
```
Comment (atumanov) [inline site/_posts/2017-11-30-ray-0.3-release.markdown]: I'm struggling with this, maybe I have it backwards, but don't you mean "Initial RLlib support for PyTorch"? AS in, "RLlib now supports PyTorch".

File: site/_posts/2017-11-30-ray-0.3-release.markdown [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+A3C in RLlib. You can try this out with the following A3C config.
+
+```bash
+cd ray/python/ray/rllib
+./train.py --run=A3C \
+           --env=PongDeterministic-v4 \
+           --config='{"use_pytorch": true, "num_workers": 8, "use_lstm": false, "model": {"grayscale": true, "zero_mean": false, "dim": 80, "channel_major": true}}'
+```
+
+## Distributed Actor Handles
```
Comment (atumanov) [inline site/_posts/2017-11-30-ray-0.3-release.markdown]: I don't like this name, but I don't hate it either. Just to capture the channel discussion, my preference would be "Distributed Actor RPC" or "Distributed Actor RMI". Up to you.

File: site/_posts/2017-11-30-ray-0.3-release.markdown [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+# Pass the actor handle to some tasks.
+tasks = [task.remote(logger, i) for i in range(10)]
+# Wait for the tasks to finish.
+ray.get(tasks)
+# Get the logs from the tasks.
+print(ray.get(logger.read_logs.remote()))
+```
+
+This feature is still considered experimental, but we've already found actor
+handles useful for implementing [**parameter server**][10] and [**streaming
```
Comment (atumanov) [inline site/_posts/2017-11-30-ray-0.3-release.markdown]: we've already found **distributed actor handles** useful for implementing...

File: site/_posts/2017-11-30-ray-0.3-release.markdown (hunk 7 above)
Comment (robertnishihara) [reply site/_posts/2017-11-30-ray-0.3-release.markdown]: Fixed

File: site/_posts/2017-11-30-ray-0.3-release.markdown (hunk 10 above)
Comment (robertnishihara) [reply site/_posts/2017-11-30-ray-0.3-release.markdown]: Fixed

**Review by atumanov — APPROVED** [review]
LGTM
